### DISCUSSION TITLE
Feature Request - Groups for Shared Expressions

### DISCUSSION THREAD

**Discussion by edhans** [TabularEditor/TabularEditor3#1139] [Ideas]
The folders/groups that we can organize M code in Power Query is available in the metadata of the shared expression for those queries that are unloaded. It would be nice if they were organized in Groups in the TOM view. This is in my Source Query group. Handy when tweaking code and working change through several stages of a query.

![image](https://user-images.githubusercontent.com/45298358/208220067-76eac8a0-9755-4f4f-98cb-da866b0c5f99.png)

**Comment by otykier** [comment]
This has been discussed to some extent in #436. The problem is that you can have M queries on partitions as well as on shared expression objects. Coming from an AS world, we prefer to let the TOM Explorer show objects in the same structure as they have in TOM, with the obvious exception of display folders for measures, columns and hierarchies.

However, in Power Query, the query groups apply to all M expressions, regardless of whether they originate from partitions or shared expressions.

So then the question is, how should we take that into account in Tabular Editor's TOM Explorer?

- Should we only show query groups for shared expressions, and ignore them for partitions? This could add confusion when comparing queries between Power Query and Tabular Editor, as you will not necessarily have the same number of queries displayed in each query group, since Tabular Editor will only group the Shared Expression queries, and not the queries originating from partitions.
- Should we get rid of the "Partitions" folder within each table, and instead combine "Shared Expressions" and partitions into a new root-level folder called "Queries", like you have in Power Query? People coming from an AS background probably won't be happy about this, as we tend to think of partitions as belonging to a table (which is also the case according to TOM...)

So the general problem is that Microsoft has defined an object model (TOM) where specific objects (M queries) are allowed to live in two very different places... what is the best way to visualize that?

**Comment by edhans** [comment]
Interesting issues. 

I had not considered some of those issues for these reasons:

- I never create another query (shared expression or partition) from a loaded query/partition. If I need to do that, I create a new unloaded query/shared expression that has all of the M from the partition M code, then change the M for that partition to just point to the new shared expression. Then I reference the next partition or shared expression from the new shared expression. I've avoided refresh failures by doing this. In other words, for me any partition is the end of the line. Nothing in M is dependent on it. Ever.
- My folder structure tends to be Source Queries, Intermediate Queries, and Reporting Queries, the latter being the only ones that are partitions. That means for me, whatever loads is also what is in the data model, plus whatever I've created _in_ the data model, like Field Parameters, Calculation Groups, or the occasional calculated table.

Coming from the Power Query/Power BI side of things, I'd side with the AS group - loaded queries should be partitions in the table they are in and in the Tables section of the TOM.

For me, the folders would _only_ be implemented in the Shared Expression section to help clean it up. I don't think there would be tremendous pushback on this. In Power BI Desktop, whatever is loaded (partitions) it is just dumped into the Field pane as a bunch of tables in alphabetical order and any folder structure they came from in Power Query is ignored. Tabular Editor should, IMHO, follow that same logic.

**Comment by otykier** [comment]
Good points. I guess grouping Shared Expressions by their query group would definitely help, even though we don't show the partition-specific queries.

**Comment by MuddasarM** [comment]
@otykier I wished for this today while working in TE3, and found it has been raised back in 2022 by several people. How can we get this into the product please?
